File: frontend/src/app/core/models/placement.ts
export type PlacementSkill = 'vocabulary' | 'grammar' | 'listening' | 'pronunciation';

export interface PlacementOption { id: string; text: string; }

export interface PlacementQuestion {
  id: string;
  skill: PlacementSkill;
  level: number;
  prompt_vi: string;
  options: PlacementOption[];
  target_text: string | null;
  target_pinyin: string | null;
  number: number;
  total: 20;
}

export interface PlacementSkillResult {
  skill: PlacementSkill;
  estimated_level: number;
  correct: number;
  assessed: number;
}

export interface PlacementResult {
  recommended_level: number;
  confidence: 'low' | 'medium' | 'high';
  confidence_vi: string;
  summary_vi: string;
  skills: PlacementSkillResult[];
  completed_at: string;
  advisory_only: boolean;
  disclaimer_vi: string;
}

export interface PlacementAttempt {
  attempt_id: string;
  status: 'in_progress' | 'completed';
  question: PlacementQuestion | null;
  result: PlacementResult | null;
}

export interface PlacementStatus {
  can_take: boolean;
  in_progress: PlacementAttempt | null;
  latest_result: PlacementResult | null;
  retake_available_at: string | null;
  selected_level: number | null;
  can_apply_level: boolean;
}

File: backend/hsk_api/models/placement.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


PlacementSkill = Literal["vocabulary", "grammar", "listening", "pronunciation"]


class PlacementOption(BaseModel):
    id: str
    text: str


class PlacementQuestion(BaseModel):
    id: str
    skill: PlacementSkill
    level: int = Field(ge=1, le=6)
    prompt_vi: str
    options: list[PlacementOption] = Field(default_factory=list)
    target_text: str | None = None
    target_pinyin: str | None = None
    number: int = Field(ge=1, le=20)
    total: int = 20


class PlacementAnswerRequest(BaseModel):
    option_id: str | None = None
    skip: bool = False


class PlacementAnswerRecord(BaseModel):
    question_id: str
    skill: PlacementSkill
    level: int
    correct: bool | None
    score: int | None = None


class PlacementSkillResult(BaseModel):
    skill: PlacementSkill
    estimated_level: int = Field(ge=1, le=6)
    correct: int = 0
    assessed: int = 0


class PlacementResult(BaseModel):
    recommended_level: int = Field(ge=1, le=6)
    confidence: Literal["low", "medium", "high"]
    confidence_vi: str
    summary_vi: str
    skills: list[PlacementSkillResult]
    completed_at: datetime
    advisory_only: bool = False
    disclaimer_vi: str = "Kết quả chỉ là gợi ý học tập, không phải điểm thi HSK chính thức."


class PlacementAttemptRecord(BaseModel):
    id: str
    account_id: str
    status: Literal["in_progress", "completed"] = "in_progress"
    target_levels: dict[str, int] = Field(default_factory=dict)
    used_question_ids: list[str] = Field(default_factory=list)
    answers: list[PlacementAnswerRecord] = Field(default_factory=list)
    current_question_id: str | None = None
    result: PlacementResult | None = None
    started_at: datetime
    completed_at: datetime | None = None


class PlacementAttemptResponse(BaseModel):
    attempt_id: str
    status: Literal["in_progress", "completed"]
    question: PlacementQuestion | None = None
    result: PlacementResult | None = None


class PlacementStatusResponse(BaseModel):
    can_take: bool
    in_progress: PlacementAttemptResponse | None = None
    latest_result: PlacementResult | None = None
    retake_available_at: datetime | None = None
    selected_level: int | None = None
    can_apply_level: bool = True


class PlacementSelectionRequest(BaseModel):
    selected_level: int = Field(ge=1, le=6)


class PlacementSelectionResponse(BaseModel):
    selected_level: int
    applied: bool = True
